lib/src/rules/declare_const_constructor_when_possible.dart

```diff
@@ -0,0 +1,111 @@
+// Copyright (c) 2017, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+library linter.src.rules.declare_const_constructor_when_possible;
+
+import 'package:analyzer/dart/ast/ast.dart';
+import 'package:analyzer/dart/ast/visitor.dart';
+import 'package:analyzer/dart/element/element.dart';
+import 'package:linter/src/analyzer.dart';
+
+const _desc =
+    r'Consider making your constructor const if the class supports it.';
+
+const _details = r'''
+
+**CONSIDER** making your constructor const if the class supports it.
+
+**BAD:**
+```
+class Bad {
+  final int foo;
+
+  Bad() : foo = 1; // LINT
+}
+```
+
+**GOOD:**
+```
+class Bad2 {
+  final int foo;
+
+  const Bad2() : foo = 1; // LINT
+}
+```
+
+''';
+
+ConstructorDeclaration _AsConstructorDeclaration(ClassMember node) {
+  if (node is ConstructorDeclaration) {
+    return node;
+  }
+  return null;
+}
+
+bool _doesNotHaveNonConstInitializer(ConstructorDeclaration element) {
+  return element.initializers.every(_isConstInitializer);
+}
+
+bool _hasEmptyBody(ConstructorDeclaration node) {
+  final body = node.body;
+  return body is EmptyFunctionBody ||
+      (body is BlockFunctionBody && body.block.statements.length == 0);
+}
+
+bool _hasNonFinalFields(ClassElement classElement) {
+  if (classElement == null) return false;
+  final superClass = classElement.allSupertypes.isNotEmpty
+      ? classElement.allSupertypes.first.element
+      : null;
+  return classElement != null &&
+      (classElement.fields.any((e) => !e.isFinal) ||
+          _hasNonFinalFields(superClass));
+}
+
+bool _isConstInitializer(ConstructorInitializer element) {
+  if (element is RedirectingConstructorInvocation) {
+    return element.staticElement?.isConst;
+  } else if (element is SuperConstructorInvocation) {
+    return element.staticElement?.isConst;
+  }
+  return true;
```

Comment:
This is incomplete in two ways. First, the spec says

> The superinitializer that appears, explicitly or implicitly, in the initializer list of a constant constructor must specify a constant constructor of the superclass of the immediately enclosing class or a compile-time error occurs.

This is missing the case where there is an implicit superinitializer. Second, the spec says

> Any expression that appears within the initializer list of a constant constructor must be a potentially constant expression ...

So if the initializer is neither of the above cases, we need to determine whether it is a potentially constant expression.